Fixed memory unsafe cast safercpp errors in UserMediaController.h

Phinny01 opened the pull request:
#### 2ee8608c8ac2db628ef036be40dc9bc9937ff22d
<pre>
Fixed memory unsafe cast safercpp errors in UserMediaController.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=296013">https://bugs.webkit.org/show_bug.cgi?id=296013</a>
<a href="https://rdar.apple.com/155914874">rdar://155914874</a>

Reviewed by Eric Carlson and Rupin Mittal.

Fix unsafe cast warning.

Source/WebCore/Modules/mediastream/UserMediaController.h
Source/WebCore/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations
Source/WebCore/platform/Supplementable.h

Canonical link: <a href="https://commits.webkit.org/297801@main">https://commits.webkit.org/297801@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ea4efd023fe8a0e4321e5be4608c780e6a4e367

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112880 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32615 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23093 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119084 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63410 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114842 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33267 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41178 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85906 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/36572 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115827 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26552 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101541 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66211 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25841 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19673 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62842 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95946 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19747 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122305 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39958 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29795 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94774 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40342 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97761 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94512 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24133 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39638 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17449 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/36071 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39844 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45342 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39484 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42817 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41222 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/mediastream/UserMediaController.h:
```diff
@@ -64,7 +64,14 @@ class UserMediaController : public Supplement<Page> {
+            return nullptr;
+
+        auto* supplement = Supplement<Page>::from(page, supplementName());
+        return supplement ? static_cast<UserMediaController*>(supplement) : nullptr;
```
this is unnecessary, if supplement is null, static_cast will also return null.

This is no different to doing:
```
return static_cast<UserMediaController*>(Supplement<Page>::from(page, supplementName()));
```

in practice your change is a no-op, from is never called with `nullptr` may as well just assert.

Phinny01 replied:
I have corrected it now

RupinMittal commented on Source/WebCore/Modules/mediastream/UserMediaController.h:
```diff
@@ -64,7 +64,14 @@ class UserMediaController : public Supplement<Page> {
+
+        auto* supplement = Supplement<Page>::from(page, supplementName());
+        return supplement ? static_cast<UserMediaController*>(supplement) : nullptr;
+    }
```
Hello! Welcome to the team!

You had a good idea, but as Jean-Yves explained above, it's not fully correct.

We fix unsafe cast Safer CPP errors by changing the `static_cast` to a `downcast`. This is a custom WebKit cast meant specifically for casting from a base class to a derived class. 

So you can fix this warning by just doing this:

```
static UserMediaController* from(Page* page) { return downcast<UserMediaController>(Supplement<Page>::from(page, supplementName())); }
```

If you make just this change and nothing else, you'll notice that you get a compiler error. It'll complain about a missing traits specialization. That's because `downcast` is a templated function and due to it's implementation (See `TypeCasts.h` if you're curious, but you don't have to), requires an implementation of the `isType()` function so that it can check if the cast it's about to do is safe. 

So you'll have to add this traits specialization. For an example for how to do this in the case when downcasting from a Supplement<T> to a derived class, see https://github.com/WebKit/WebKit/pull/47832

Hope this helps!

Phinny01 replied:
Hello! Thank you. This was really helpful. I think I have addressed the issues now.

RupinMittal commented on Source/WebCore/Modules/mediastream/UserMediaController.h:
```diff
@@ -113,4 +112,8 @@ inline void UserMediaController::updateCaptureState(const Document& document, bo
 
+SPECIALIZE_TYPE_TRAITS_BEGIN(WebCore::UserMediaController)
+    static bool isType(const WebCore::Supplement<WebCore::Page>& supplement) { return is<WebCore::UserMediaController>(supplement); }
+SPECIALIZE_TYPE_TRAITS_END()
```
Wow! I'm simultaneously amazed and also worried that this even compiled on Mac. Perhaps the compiler we use on Windows is better. 

The issue with your code is that it implements the `is()` function in terms of itself (AKA infinite recursion). 

What we want to do instead is add a new function that determines if the supplement is a `UserMediaController` and use that instead:

```
static bool isType(const WebCore::SupplementBase& supplement) { return supplement.isUserMediaController(); }
```

This new `isUserMediaController()` function should be virtual and return false in the base class (`SupplementBase`) and be overridden and return true in the derived class (`UserMediaController`). 

WebKit code can be quite confusing at times--especially with all the templates--but if you follow the example in https://github.com/WebKit/WebKit/pull/47832, it should work :)

RupinMittal commented on Source/WebCore/Modules/mediastream/UserMediaController.h:
```diff
@@ -64,9 +64,13 @@ class UserMediaController : public Supplement<Page> {
+    static UserMediaController* from(Page* page)
+    {
+        return downcast<UserMediaController>(Supplement<Page>::from(page, supplementName()));
+    }
```
Let's keep this function on one line as it was before.

RupinMittal commented on Source/WebCore/Modules/mediastream/UserMediaController.h:
```diff
@@ -113,4 +116,8 @@ inline void UserMediaController::updateCaptureState(const Document& document, bo
 } // namespace WebCore
 
+SPECIALIZE_TYPE_TRAITS_BEGIN(WebCore::UserMediaController)
+    static bool isType(const WebCore::Supplement<WebCore::Page>& supplement) { return supplement.isUserMediaController(); }
```
Let's have the function take in `const WebCore::SupplementBase& supplement` as mentioned in the previous comment.

RupinMittal commented on Source/WebCore/platform/Supplementable.h:
```diff
@@ -82,6 +82,8 @@ class SupplementBase {
     virtual bool isNavigatorClipboard() const { return false; }
     virtual bool isNavigatorCookieConsent() const { return false; }
+
+    virtual bool isUserMediaController() const { return false; }
```
Let's remove the extra line above this.

RupinMittal reviewed (commented):
Almost there. Your commit message mentions that we're fixing infinite recursion here. But we're not--there's no infinite recursion in the current code (the infinite recursion was in the code added in the previous attempt to fix in the unsafe cast issue). 

This patch is simply fixing the unsafe cast issue. So the commit message should say something about that. Doesn't have to be complicated, something simple like "Fix unsafe cast warning." is enough.

(This doesn't actually make a difference to the code, but it's good to keep our commit messages clean since having a good commit history is helpful when we find issues and have to go look at previous commits).

RupinMittal commented on Source/WebCore/Modules/mediastream/UserMediaController.h:
```diff
@@ -113,4 +114,8 @@ inline void UserMediaController::updateCaptureState(const Document& document, bo
 } // namespace WebCore
 
+SPECIALIZE_TYPE_TRAITS_BEGIN(WebCore::UserMediaController)
+    static bool isType(const WebCore::SupplementBase supplement) { return supplement.isUserMediaController(); }
```
Could we use `const WebCore::SupplementBase& supplement` here as I mentioned in my previous comments?

We're stilling missing the `&`. Thanks!

Phinny01 replied:
oh yes I have fixed that now

RupinMittal reviewed (approved):
LGTM. Let's make sure the bots are green and @jyavenard approves before landing.

Commit:
Fixed memory unsafe cast safercpp errors in UserMediaController.h
https://bugs.webkit.org/show_bug.cgi?id=296013
rdar://155914874

Reviewed by Eric Carlson and Rupin Mittal.

Fix unsafe cast warning.

Source/WebCore/Modules/mediastream/UserMediaController.h
Source/WebCore/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations
Source/WebCore/platform/Supplementable.h

Canonical link: https://commits.webkit.org/297801@main
